sdm: add failure mode analysis (#388)
ethereum-optimism/design-docs: security/fma-sdm-postexec.md

  3. Tests cover producer and verifier behavior near the gas limit and payloads that attempt to hide pre-refund overuse.
- **Possible Mitigations:**
  1. Add an explicit end-of-block assertion that total pre-refund gas remains within the block limit.
  2. Alert on unusually high raw EVM gas relative to canonical gas.
- **Existing Detection:** Block execution time and replica lag are observable through standard node monitoring.
- **Possible Detection:** Expose raw EVM gas alongside canonical gas and alert on an unusually high ratio.
- **Recovery Path(s):** Disable SDM production or reduce sequencing load while patching the affected builder or verifier. If deployed clients disagree on the resource-limit rule, recovery follows the chain-split path in FM1b.

### FM4: Fault-proof programs cannot prove an SDM block

- **Description:** An execution client may validate an SDM block while the deployed fault-proof program fails to reproduce it. Causes include different transaction decoding, settlement behavior, configuration, or behavior that works in native execution but fails inside the production FPVM. The deployed proof program and prestate must support Lagoon and SDM.

> - **Risk Assessment:** High impact, medium likelihood until production-FPVM coverage exists. An unprovable output stalls permissionless withdrawals and may require replacing the fault-proof program under governance.

Author: > no native/FPVM divergence has been found or ruled out, right?

No, and I wouldn't expect to find one. If we have tests for SDM with native execution in `kona` (and we do), that's good enough, and we don't need any more tests.

We shouldn't have to run the compiled Kona program inside the production Cannon FPVM.

- **Risk Assessment:** Medium impact, low likelihood with configuration monitoring. Drift does not invalidate blocks or halt the chain, but it can create inconsistent user charges, revenue, and operator-policy compliance.
- **Existing Mitigations:**
  1. Refund production is enabled only when both Lagoon is active and the operator opt-in is set.
  2. Verification ignores the local production opt-in and does not recompute the sequencer's refund policy.
  3. The operator opt-in defaults off, supports boot-time configuration, and exposes effective state through the admin status method.
- **Possible Mitigations:**
  1. Add restart and cross-sequencer health checks that verify the intended effective opt-in.

>   2. Document and test that mutating admin methods are not exposed on public HTTP or WebSocket endpoints.

Reviewer: The mutating admin method is not persistent either, right? It would be good to:
1. Persist admin mutation state
2. Remove persisted admin mutation state on boot if it matches boot config
3. Alert when different than boot config state

This combination makes it so the mitigation doesn't resolve alerts until it is addressed completely (i.e. update the config and restart). This could be a good larger-scale operational feature than SDM, though. Might suggest mitigating via clearly-defined runbook and defer this boot config vs run config discrepancy detection, alerting, etc.

Author: I don't think we have any code that persists configs / admin mutations. I am reluctant to introduce this as part of SDM honestly.

compatibility.

SDM lets the sequencer reduce the gas charged to a standard Ethereum transaction. The sequencer appends at most
one `0x7D` transaction as the final transaction in a block. Its payload contains the block number and a list of
`(transaction index, gas refund)` entries. Verifiers apply the supplied values when calculating canonical gas,
receipts, fee settlement, and the state root.

The refund **mechanism** is consensus-critical, but the refund **policy** is sequencer-defined. Verifiers enforce
payload structure, `refund <= evmGasUsed`, and settlement validity; they do not recompute whether a refund was
justified by the sequencer's intended policy. The version-1 production policy is block-level warming: it rebates
an EIP-2929 cold-to-warm surcharge only when the transaction actually paid that surcharge. This policy boundary
is an explicit sequencer trust assumption.

SDM activates with the Lagoon network upgrade. Producing refunds additionally requires an operator-controlled
execution-layer opt-in. Verification does not depend on this opt-in: after activation, every verifier must accept

> and apply a valid SDM payload whether or not that verifier would produce one.

Reviewer 2: ```suggestion
and apply a valid SDM payload if the sequencer did.
```

> and apply a valid SDM payload whether or not that verifier would produce one.

Reviewer 3: ```suggestion
and apply a valid SDM payload.
```

The normative specifications are the source of truth:

> - [Post-Execution Transactions](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/post-exec.md)

Reviewer 2: Do we have specs covering the behaviour of SDM as it relates to subblocks?

This document implies there are specific choices made about how it should work, so we ought to have a spec for it.

Author: Not currently, the subblocks behavior is described only in the SDM design doc. I am not sure there is another way it could work though, given the requirement that every subblock should be a valid block.

Author: Added as a proposed item, so we can discuss and make a decision.

Author: @geoknee on a second thought - I think this is really an implementation detail, and not something we should have specs for, but I am open to discuss it further.

Invariants are:
- every subblock must produce a valid block
- PostExec tx is accumulating refunds

Then clearly we have a `replace-not-append` semantics.

- [Sequencer-Defined Metering](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/sdm.md)
- [Lagoon Network Upgrade](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/overview.md)
- [SDM design document](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/sdm.md)

## Failure Modes and Recovery Paths

> ### FM1: PostExec construction or application diverges, causing invalid blocks or chain splits

Reviewer 2: Might be worth splitting invalid blocks and chain splits into an FM1a and FM1b, since the recovery path may be quite different in either case.

Author: Agreed.

> ### FM1: PostExec construction or application diverges, causing invalid blocks or chain splits

Reviewer 3: This FM could be **structurally invalid blocks**.

Author: Agreed. This is now FM1a: “Structurally invalid or inconsistent PostExec blocks.”

- **Description:** The producer's streamed, derived, or sealed block views may contain different PostExec

>   payloads; a sealed payload may be malformed; or consensus consumers may apply the same valid payload

Reviewer 3: what exactly is a 'block view'. And in which sense 'may it contain **different** PostExec payloads'.

Author: The ambiguity has been removed.

  transactions, incorrect position or block anchor, invalid transaction indices, applying a refund twice, or
  calculating settlement differently. A stream-only mismatch breaks pending views and preconfirmations. Uniform
  rejection of a malformed sealed block causes a deposits-only replacement and user-transaction reorg. A
  validity or state-root disagreement between clients causes a chain split.
- **Risk Assessment:** High impact, low likelihood. PostExec affects consensus-visible gas accounting, receipts,
  balances, and state roots. The specification, shared execution code, and execution-layer validation reduce the
  likelihood, but producer lifecycle bugs or an independent implementation can still diverge.
- **Existing Mitigations:**
  1. The specifications require at most one final `0x7D`, the correct block-number anchor, an active schema, and
     non-empty, non-zero entries ordered by strictly increasing transaction index.
  2. Entries may target only standard Ethereum transactions. Execution enforces `refund <= evmGasUsed`, rejects
     settlement underflow, and consumes every entry exactly once.

>   3. Incremental streams use replace-not-append semantics for the cumulative PostExec payload; the final streamed

Reviewer 2: We use "streams" quite a bit, and I guess we mean the subblocks stream. Better to just use that concrete identifier IMO.

Author: Yes, streams here is referencing subblocks.

     bytes are compared with the sealed transaction, and sealing aborts if PostExec finalization fails.

>   4. Derivation enforces activation while the execution layer remains authoritative for post-activation payload

Reviewer 2: In what sense does "derivation enforce activation:?

     validity and deposits-only recovery.
  5. Existing unit and acceptance tests cover structural rejection, producer/verifier round trips, and
     streamed-to-sealed payload equality.
- **Possible Mitigations:**

>   1. Add a non-empty fixture that compares execution-client and fault-proof-program block hash, receipts root,

Reviewer 2: We can probably just say "check for consensus between" rather than listing all of these fields. If the block hash agrees, this implies the other fields do too.

What is the "non empty fixture"? Would this be the action tests that we decided not to do in the end?

Author: Agreed. “Non-empty” means a block containing at least one SDM refund.

     gas used, and state root.

>   2. Add deposits-only replacement metrics and cross-client malformed-payload recovery coverage.

Reviewer 2: I believe we will only get a deposit-only replacement once an invalid input is batched to L1 and then derived. We would most likely want to (and maybe already would with our current alerts) learn of a problem much sooner, when verifiers try to insert the invalid unsafe payload.

Reviewer 2: So we will see an unsafe head stall, and if it can be attributed to an invalid unsafe payload caused by SDM, I would suggest the runbook would have us:
* stop the batcher to (maybe, if done quickly enough) prevent the invalid data being batched to L1
* hit the kill switch to disable SDM (we now suspect a bug in the sequencer or verifier code or both)
* rewind the sequencer to the block before the bad one

>   3. Add consumer-side checks and alerts for missing or inconsistent cumulative streamed payloads.

Reviewer 3: I think we have logging of most errors, but what about high error rates - like if 90% of blocks fail to validate  - do we get notified somehow?

Reviewer 3: **Opus 5 (high)**: No — today nothing counts a rejection, so there is no series to alert on.

The SDM dashboard (`optimism-premium/observability/sdm/dashboard.json`) has post-exec panels, but they're built on `reth_sync_execution_post_execution_histogram_count` and `reth_sync_block_validation_post_execution_validation_duration_count` — validations *performed* and how long they took — plus `reth_payloads_failed_jobs` on the build side. There is no `..._rejected_total` / `..._invalid_total`, and no alert rule in `k8s/grafana-cloud/terraform-rules/` mentions sdm, postexec, or refund (only `subblocks.yaml`, 4 alerts).

What would actually catch the 90% case, and only as a second-order symptom:

- **uniform rejection** → chain stops advancing → head-lag and `sequencing-window-expiry-risk` alerts in `replicas.yaml`. Loud, but says nothing about the cause.
- **partial rejection / client split** → `healthcheck_is_currently_diverged` (P1, 5m) and `replica-target-height-low-match`.
- **90% failing while the chain still advances** — e.g. the stream-only mismatches that self-heal per this FM's recovery path — → nothing. Logs only, and nobody pages on log volume.

The precedent to copy is right next door: subblocks shipped with counter-rate alerts (`increase(reth_subblocks_broadcast_success_total[5m]) < 1`, dropped-subblock counters summed over 5m, p5-vs-target cadence). So I'd turn the two "Possible" bullets into a concrete ask: `postexec_validation_rejected_total` labeled by reason (structural / entry / settlement), `deposits_only_replacement_total`, an alert on rejected-over-total exceeding a threshold for 5m, and a page on any nonzero deposits-only replacement — as a pre-mainnet-activation requirement rather than an optional mitigation.

Worth also rewording **Existing Detection**: "execution-engine `INVALID` responses" reads like alerting when it's really just logs.

Author: If 90% of blocks fail to validate for whatever reason, this is out of scope for SDM, we should have alerts on ERROR rates and second-order effects for that.

  choice of canonical behavior.

### FM2: Incorrect refunds cause economic loss

- **Description:** A block may be structurally valid while carrying refunds that do not match the operator's
  intended policy. An over-refund reduces fee-vault or sequencer revenue; an under-refund overcharges users; a
  selectively applied policy may violate operator commitments. Verifiers cannot distinguish an intended refund
  from an erroneous one because refund policy is not part of consensus. Historically, this class included
  rebates for intrinsically warm or otherwise unmetered accesses: the transaction received a cold-to-warm rebate
  even though the EVM never charged the cold surcharge.
- **Risk Assessment:** Medium impact, low likelihood. The likely outcome is bounded economic misallocation rather
  than a chain split. A settlement conservation bug would have higher impact, but is constrained by consensus
  checks and conservation tests.
- **Existing Mitigations:**
  1. Consensus limits each refund to the transaction's `evmGasUsed` and rejects settlement underflow.
  2. Settlement tests assert that sender credits equal recipient debits across transaction types, fee settings,

>      and arithmetic boundaries.

Reviewer 2: This feels like a mitigation for an infinite ETH mint bug or similar, rather than economic loss for the operator. I suggest that should be it's own failure mode (very high impact, very low probability).

Reviewer 2: There's also "draining the fee vaults". Not as bad as infinite ETH mint bug, but worse than simply "lost revenue".

Author: Agreed.

  3. Policy tests cover EIP-2929 metered accesses and known intrinsically warm or protocol-generated accesses that
     must not receive a rebate.
  4. Stateful policy data is snapshotted and restored when a candidate transaction fails execution or is not
     committed to the block.
  5. Operators can disable refund production without changing verification behavior.

> - **Possible Mitigations:**

Reviewer 3: What if we had policy-combinators, e.g.

    `CeilPolicy(FaultyBlockWarmingPolicy, PercentagePolicy(50))`

This way we could bound the error and risk.

Author: This would add configuration and audit complexity, so we'd rather not do it for now.

  1. Independently audit the production policy and its block-building integration.
  2. Add sampled policy-aware reconciliation and alerts for anomalous refund volume or refund-to-raw-gas ratios.
- **Existing Detection:** Refund totals and fee-vault balance changes are observable through metrics and
  canonical block data.
- **Possible Detection:** Alert on refund-to-raw-gas anomalies and compare sampled payloads with an independent
  policy-aware checker.
- **Recovery Path(s):** Disable refund production, correct the policy, and redeploy. Already finalized refunds are
  consensus-valid and should not be rolled back; material losses require accounting and, if appropriate,
  treasury reconciliation. A true value-conservation failure requires an emergency upgrade.

### FM3: Refunds bypass block resource limits

- **Description:** Canonical gas decreases after refunds. If block admission used only post-refund gas, a producer
  could include substantially more EVM work than the block gas limit intends, causing replicas and challengers to

>   fall behind even though the header's `gasUsed` remains low.

Reviewer 2: This is essentially a Denial Of Service vector?

Author: Yes, I'll make that explicit.

- **Recovery Path(s):** Disable SDM production or reduce sequencing load while patching the affected builder or
  verifier. If deployed clients disagree on the resource-limit rule, recovery follows the chain-split path in
  FM1.

### FM4: Fault-proof programs cannot prove an SDM block

- **Description:** An execution client may validate an SDM block while the deployed fault-proof program fails to
  reproduce it. Causes include different transaction decoding, settlement behavior, configuration, or behavior
  that works in native execution but fails inside the production FPVM. SDM also activates with Lagoon, whose
  dependency-set and proof-program requirements must be satisfied by the target chain.
- **Risk Assessment:** High impact, medium likelihood until production-FPVM coverage exists. An unprovable output
  stalls permissionless withdrawals and may require replacing the fault-proof program under governance.
- **Existing Mitigations:**
  1. Fault-proof execution uses the same PostExec parser and refund application logic as the execution client.
  2. Native proof tests exercise non-empty SDM blocks.
- **Possible Mitigations:**

>   1. Prove a non-empty SDM block in the production FPVM rather than only running the program natively.

Reviewer 2: We should check in with a proofs expert about what our standards are on this kind of thing. IIUC this can take a long time to run through (playing a dispute game down to a single step and executing that on chain). 

My assumption is that we usually rely on testing equivalence between the client code and the fault proof _program_, and take the virtual machine part and test that separately.

  it inactive. Legacy configuration keys may be silently ignored. An execution client without PostExec support
  cannot follow an SDM-active chain. Missing proof-program prerequisites can prevent proving at activation.
  Separately, the operator opt-in defaults off and may differ across sequencer instances, causing refunds to stop
  or vary after a restart. Exposing the mutating admin method can let an unauthorized caller toggle production.
- **Risk Assessment:** High impact, low likelihood with an activation preflight. Cross-component activation
  disagreement can halt or split the chain. An opt-in-only mismatch has lower impact because verification is
  independent of opt-in and missing refunds remain valid.
- **Existing Mitigations:**
  1. Every component derives SDM activation from the Lagoon schedule.
  2. Refund production is enabled only when both Lagoon is active and the operator opt-in is set.
  3. Verification ignores the local production opt-in.
  4. The operator opt-in defaults off, supports boot-time configuration, and exposes effective state through the
     admin status method.

> - **Possible Mitigations:**

Reviewer 2: We should probably mention the superchain registry here, that is the main tool we use to help ensure all nodes have the correct timestamp for a hardfork. 

This would also be a good opportunity to refresh and reference https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-generic-hardfork.md since it is a general concern and not special to SDM.

Author: Agreed. I’ll add the Superchain Registry as the expected activation schedule. The generic hardfork FMA is already referenced.

  1. Add a cross-component activation preflight covering timestamps, legacy keys, execution-client support,
     proof-program selection, and dependency-set configuration.
  2. Add restart health checks that verify the intended effective opt-in state.

>   3. Document and test that mutating admin methods are not exposed on public HTTP or WebSocket endpoints.

Reviewer 3: Do we have a mechanism like a 24hr ahead-of-launch NOOP log statement, where nodes state their intent to activate in 24 hrs in a way that we can verify?

Reviewer 3: **Opus 5 (high)**: There's no NOOP-log or attestation mechanism, and I don't think we need to invent one — the intent is already queryable per node, ahead of time, in three places:

- **`admin_sdmStatus(query_timestamp)`** — on develop in both op-reth (`rust/op-reth/crates/rpc/src/sdm_admin.rs`) and op-rbuilder. It returns `operatorSdmOptIn`, `protocolActive` evaluated at *any* timestamp you pass, `effective`, and `activationTime`. Passing `activationTime + 1` is literally "do you intend to be SDM-active then?", 24h out or 24 days.
- **`eth_config`** (EIP-7910), op-geth `internal/ethapi/api.go:1422` — `current`/`next`/`last`, each with `activationTime` and `forkId`. OP fork timestamps are `*Time` fields on `ChainConfig` and `gatherForks` collects them reflectively, so the SDM gate is inside that `forkId`.
- **`optimism_rollupConfig`** (`op-node/node/api.go:175`) for the CL-side schedule, with the superchain registry as the expected value.

So the two real gaps, which is what I'd add to the doc:

1. **Nothing scrapes or compares any of it.** Verification is manual. Possible Mitigation 1 should say the preflight is a *scheduled* check that diffs each node's `next.activationTime`/`forkId` and `admin_sdmStatus(T+1)` against the registry starting N hours out, and pages on disagreement. That's the checkable form of the 24h intent statement.
2. **`admin_sdmStatus` can't be the fleet-wide signal.** It's admin-namespace, and mitigation 3 on this very line says keep admin off public endpoints — so it's unavailable for third-party nodes and isn't Prometheus-scraped. Cheapest fix: expose the same three values as gauges (`sdm_protocol_active`, `sdm_operator_opt_in`, `sdm_activation_time`). Metrics already flow fleet-wide through existing scrape config, which makes a cross-node disagreement alert a one-liner *and* makes it work before activation.

One asymmetry worth noting: op-node has no equivalent status method on develop (the `SdmStatus` doc comment claims the shape matches op-node, but nothing there implements it), so CL-side intent is only inferable from `optimism_rollupConfig`. At the protocol layer, EIP-2124 forkid does carry the next scheduled fork in the EL handshake, so ELs can in principle detect future-fork disagreement pairwise at peering time — but that covers neither the CL nor the operator opt-in.

  Strict consumers may reject the unfamiliar `0x7D` shape, assume every transaction has signature fields, or
  mishandle its zero-gas receipt. Standard transaction receipts additionally expose `opGasRefund`, indexed by the
  transaction's block-global position. Filtering or reordering transactions can attribute a valid refund to the
  wrong receipt. Inconsistent auxiliary fee fields on the PostExec receipt can also confuse explorers and
  accounting systems without affecting consensus.
- **Risk Assessment:** Medium impact, low-to-medium likelihood. This is primarily an ecosystem compatibility and
  accounting risk, not a consensus risk.
- **Existing Mitigations:**
  1. The PostExec specification defines the minimal RPC transaction shape, receipt behavior, and
     `opGasRefund` semantics.
  2. Payload validation guarantees ordered, unique, in-range indices before receipt projection.
  3. Unit and acceptance tests cover PostExec serialization and receipt refund projection.
- **Possible Mitigations:**
  1. Add end-to-end coverage for `eth_getBlockByNumber`, `eth_getTransactionByHash`, transaction/receipt
     cross-links, and representative generic Ethereum decoders.

>   2. Publish the RPC schema and coordinate with explorer and infrastructure providers before activation.

Reviewer 2: We could add here, as discussed, exposing a public SDM-enabled devnet for ecosystem integrators to build against. Base do something similar and call it a "vibe net". Or simply pointing integrators to a tool like kupcake so they can run a devnet like that themselves locally.

Author: > exposing a public SDM-enabled devnet for ecosystem integrators to build against

I'd rather we don't do any of these, until at least one ecosystem integrator has requested anything like this.

Ecosystem integrators just need to be aware of the fact that 0x7d will exist, and show its contents, nothing more is required of them - there is no need for any special visualization or parsing, unless we want first-class support for showing tuples (tx-id, refund).

We don't do anything special for the Deposit tx either -- here is an example: https://optimistic.etherscan.io/tx/0x9d2928961ac70c8c195200cf0d74a985b376c1b2300312bbb5c0c0442f94553d

  integration testing.
- **Recovery Path(s):** Patch the RPC serialization or consumer integration. Historical refund data remains
  recoverable from the canonical PostExec payload. RPC-only fixes do not require a hardfork unless the underlying
  consensus receipt is incorrect.

### Generic items

The [generic hardfork failure modes](https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-generic-hardfork.md)
apply, especially activation coordination, cross-client divergence, chain-split detection, fault-proof prestate
updates, and rollback procedures. SDM adds no L1 or L2 contracts, so the generic smart-contract FMA is not
applicable.

- [x] The applicable generic hardfork failure modes have been reviewed and incorporated above.

## Action Items

The following are candidate follow-ups for Security and engineering review. They are **not commitments or launch
blockers** until an owner accepts them and a tracking issue or PR is created.

| ID | Candidate Follow-up | Failure Modes | Owner | Status |
| --- | --- | --- | --- | --- |
| A1 | Run a production-FPVM proof over a block with a non-empty SDM payload. | FM1, FM4 | _TBD_ | Proposed |

> | A2 | Add a post-Isthmus, Lagoon-active, non-empty fixture that compares execution-client and proof-program block hash and state root. | FM1, FM4 | _TBD_ | Proposed |

Reviewer 2: ```suggestion
| A2 | Add an action test that compares execution-client and proof-program block hash and state root. | FM1, FM4 | _TBD_ | Proposed |
```
Here's the kicker: this reads like overly-verbose, AI generated material. Let's be more terse.

> | A3 | Document the proof-program and dependency-set requirements for every SDM-active chain. | FM4, FM5 | _TBD_ | Proposed |

Reviewer 2: Aren't dependency-set requirements being documented a concern for the interop team / interop FMA?

Author: Yes, you're right. This was an issue when starting a network with SDM, as SDM is ran on Lagoon, but really this is out of scope for SDM.

| A9 | Add consumer-side cumulative-payload checks and pending-versus-sealed mismatch alerts. | FM1 | _TBD_ | Proposed |
| A10 | Add refund anomaly alerts and sampled policy-aware reconciliation. | FM2, FM3 | _TBD_ | Proposed |
| A11 | Decide and document the PostExec receipt's auxiliary L1 fee fields. | FM6 | _TBD_ | Proposed |
| A12 | Add an operator runbook for opt-in health checks, restarts, and admin-RPC exposure. | FM5 | _TBD_ | Proposed |

- [ ] Resolve all review comments and incorporate accepted decisions into this document (Assignee: document author).

## Audit Requirements

**Recommendation: complete an external audit before production activation.**

The audit should cover:

- PostExec decoding and structural validation;
- SDM canonical-gas calculation and fee settlement, including value conservation and arithmetic boundaries;
- pre-refund block resource-limit enforcement;
- transaction rollback and policy snapshot behavior;
- producer, streamed-subblock, and sealed-block consistency;
- production refund-policy correctness against the EIP-2929 metered-access rules; and

> - execution-client versus fault-proof-program parity, including the production FPVM.

Reviewer 2: Again I'm not sure if worrying about the FPVM here is really necessary. Let's discuss.

Author: Agreed that execution-client versus proof-program parity is the important SDM-specific property.

# Sequencer-Defined Metering (SDM) and Post-Execution Transactions: Failure Modes and Recovery Path Analysis

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Introduction](#introduction)

> - [Failure Modes and Recovery Paths](#failure-modes-and-recovery-paths)

Reviewer 4: Should we include batcher-related failure modes? Batcher needs to batch postExec txs, so this should be also covered?

Author: Agreed. Although the L1 batcher transaction format is unchanged, singular- and span-batch encoding must preserve the PostExec transaction and its ordering. Omitting it can produce a valid but different derived block without refunds, while corruption or reordering can make the derived payload invalid. I’ll incorporate this into FM1a and mention the existing singular/span derivation coverage.

justified by the sequencer's intended policy. The version-1 production policy is block-level warming: it rebates
an EIP-2929 cold-to-warm surcharge only when the transaction actually paid that surcharge. This policy boundary
is an explicit sequencer trust assumption.

SDM activates with the Lagoon network upgrade. Producing refunds additionally requires an operator-controlled
execution-layer opt-in. Verification does not depend on this opt-in: after activation, every verifier must accept
and apply a valid SDM payload if the sequencer did.

The normative specifications are the source of truth:

- [Post-Execution Transactions](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/post-exec.md)
- [Sequencer-Defined Metering](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/sdm.md)
- [Lagoon Network Upgrade](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/overview.md)
- [SDM design document](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/sdm.md)

## Failure Modes and Recovery Paths

### FM1a: Structurally invalid or inconsistent PostExec blocks

> - **Description:** A producer may emit a stale cumulative PostExec payload in a subblock stream, produce a final

Reviewer 4: Should generalize this description to also capture blocks which has subblocks disabled. SDM can be enabled even if we disable subblocks.

Author: Agree. I'll update the doc to say that malformed sealed PostExec blocks are a general risk.

  subblock-stream payload that differs from the sealed transaction, or seal a malformed payload. Examples include
  multiple `0x7D` transactions, an incorrect position or block anchor, and invalid transaction indices. A
  subblock-stream-only mismatch breaks pending views and preconfirmations. Uniform rejection of a malformed sealed
  block causes a deposits-only replacement and user-transaction reorg.
- **Risk Assessment:** High impact, low likelihood. Producer lifecycle bugs can invalidate preconfirmations or omit
  user transactions from the derived chain. The specification, execution-layer validation, and lifecycle tests
  reduce the likelihood.
- **Existing Mitigations:**
  1. The specifications require at most one final `0x7D`, the correct block-number anchor, an active schema, and
     non-empty, non-zero entries ordered by strictly increasing transaction index.

>   2. Incremental subblock streams use replace-not-append semantics for the cumulative PostExec payload; the final

Reviewer 4: Can you please elaborate more on this mitigation?

> use replace-not-append semantics

What does replace-not-append mean? 

> final subblock-stream payload is compared with the sealed transaction

Does "sealed transaction" mean the postExec tx? If yes, shouldn't it be compared with every subblock stream(0th subblock to Nth subblock)

> sealing aborts if PostExec finalization fails.

so.. what happens if abort happens? We get a canonical block which does not have postExec Tx? We cannot build a canonical block at all?

Author: Replace-not-append means every subblock carries the latest PostExec candidate separately from its transaction delta. Earlier candidates do not contain all refunds, since refunds accumulate with every new subblock.

     subblock-stream payload is compared with the sealed transaction, and sealing aborts if PostExec finalization fails.
  3. Before Lagoon activation, blocks containing PostExec transactions are invalid. After activation, the
     execution layer validates PostExec payloads, while derivation handles deposits-only recovery when a payload
     is invalid.
  4. Existing unit and acceptance tests cover structural rejection and subblock-stream-to-sealed payload equality.
- **Possible Mitigations:**
  1. Add deposits-only replacement metrics and cross-client malformed-payload recovery coverage.

> - **Existing Detection:** Execution-engine `INVALID` responses and differential-test failures.

Reviewer 4: Does exec engine throw `INVALID` when it tries to sanity check the structure enforced by https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/post-exec.md#block-level-structural-rules ?

If yes, can we also emit metrics at the exec engine to detect the reason of `INVALID` happened because of postExec structure check?

Author: Sure, will add metrics for the INVALID reasons.

- **Possible Detection:** Alerts for deposits-only replacements.
- **Recovery Path(s):** A subblock-stream-only mismatch self-heals when the canonical block arrives but invalidates
  the affected preconfirmation. Uniform rejection of a sealed payload results in a deposits-only block and
  requires users to resubmit omitted transactions.

> ### FM1b: PostExec application diverges across clients

Reviewer 4: My understanding is that this application happens at execution client, and we only maintain op-reth. So my ask is, which clients are we comparing? op-reth vs kona-executor ?

Author: Yes, the relevant comparison today is op-reth’s native execution path versus kona-executor’s fault-proof path. They share core execution code, but their parsing, configuration, and integration paths can still diverge. I’ll update the wording to make this concrete.

- **Recovery Path(s):** A live client split requires an emergency release and coordinated choice of canonical
  behavior.

### FM2a: Incorrect policy refunds cause economic loss

- **Description:** A block may be structurally valid while carrying refunds that do not match the operator's
  intended policy. An over-refund reduces fee-vault or sequencer revenue; an under-refund overcharges users; a
  selectively applied policy may violate operator commitments. Verifiers cannot distinguish an intended refund
  from an erroneous one because refund policy is not part of consensus. Historically, this class included
  rebates for intrinsically warm or otherwise unmetered accesses: the transaction received a cold-to-warm rebate
  even though the EVM never charged the cold surcharge.
- **Risk Assessment:** Medium impact, low likelihood. The economic misallocation is bounded per transaction by its
  `evmGasUsed` and does not cause a chain split.
- **Existing Mitigations:**
  1. Consensus limits each refund to the transaction's `evmGasUsed`.
  2. Policy tests cover EIP-2929 metered accesses and known intrinsically warm or protocol-generated accesses that
     must not receive a rebate.

>   3. Stateful policy data is snapshotted and restored when a candidate transaction fails execution or is not

Reviewer 4: Is this part of our spec? Also

> candidate transaction fails execution

this happen? This can happen because of incorrect policy refunds?

Author: No, we can't fail a tx because of incorrect refund.

  4. Operators can disable refund production without changing verification behavior.
- **Possible Mitigations:**
  1. Independently audit the production policy and its block-building integration.
  2. Add sampled policy-aware reconciliation and alerts for anomalous refund volume or refund-to-raw-gas ratios.
- **Existing Detection:** Refund totals and fee-vault balance changes are observable through metrics and
  canonical block data.
- **Possible Detection:** Alert on refund-to-raw-gas anomalies and compare sampled payloads with an independent
  policy-aware checker.
- **Recovery Path(s):** Disable refund production, correct the policy, and redeploy. Already finalized refunds are
  consensus-valid and should not be rolled back; material losses require accounting and, if appropriate,
  treasury reconciliation.

### FM2b: Settlement violates value conservation

- **Description:** A settlement implementation bug may credit the sender without applying equal debits, debit the
  wrong fee recipient, or otherwise violate value conservation. If deployed consistently across clients, this
  could mint ETH or drain fee-vault or beneficiary balances without causing an immediate chain split.

> - **Risk Assessment:** Very high impact, very low likelihood. A value-conservation failure can corrupt canonical

Reviewer 4: We may maintain a external monitor like monitorism / hexagate to detect that the invariant holds.

  balances, but the settlement equations, underflow checks, and conservation tests substantially reduce the
  likelihood.
- **Existing Mitigations:**
  1. Settlement requires sender credits to equal the sum of recipient debits and rejects debit underflow.
  2. Settlement tests cover transaction types, fee settings, and arithmetic boundaries.
  3. Operators can disable refund production without changing verification behavior.
- **Possible Mitigations:**
  1. Independently audit settlement and add invariant or property tests for value conservation.
  2. Reconcile sender credits and fee-recipient debits from sampled canonical blocks.
- **Existing Detection:** Conservation and differential tests detect known settlement discrepancies before
  deployment.

> - **Possible Detection:** Alert on reconciliation failures or anomalous fee-vault and beneficiary balance changes.

Reviewer 4: How will we detect this "reconciliation failures"?

Author: I agree we need such alerts. We don't have alerts right now. See https://github.com/ethereum-optimism/protocol-team/issues/237

- **Recovery Path(s):** Disable refund production immediately and deploy corrected clients. Canonical balance
  corruption may require an emergency upgrade and the established incident-response process.

### FM3: Refunds bypass block resource limits

- **Description:** Canonical gas decreases after refunds. If block admission used only post-refund gas, a producer

>   could include substantially more EVM work than the block gas limit intends, creating a resource-exhaustion

Reviewer 4: Is this only possible when the postExec tx was unintentionally inserted between user txs?

Author: I am not sure I follow. PostExec tx intentionally (or unintentionally) inserted between user tx would result into an invalid block.

  must support Lagoon and SDM.
- **Risk Assessment:** High impact, medium likelihood until production-FPVM coverage exists. An unprovable output
  stalls permissionless withdrawals and may require replacing the fault-proof program under governance.
- **Existing Mitigations:**
  1. Fault-proof execution uses the same PostExec parser and refund application logic as the execution client.
  2. Native proof tests exercise non-empty SDM blocks.
- **Possible Mitigations:**
  1. Prove a non-empty SDM block in the production FPVM rather than only running the program natively.
  2. Add a post-Isthmus, Lagoon-active fixture that compares execution-client and proof-program block hash and
     state root.
- **Existing Detection:** Proof-runner failures, challenger disagreement, and output-root disagreement in native
  acceptance tests.
- **Possible Detection:** A production-FPVM regression test that detects native-versus-FPVM trace divergence.
- **Recovery Path(s):** Deploy a corrected proof program and publish a new absolute prestate through the approved
  governance and registry process. Use the established fault-proof recovery process while the corrected program
  is prepared.

> ### FM5: Activation or operator configuration is inconsistent

Reviewer 4: I would decouple this FM5 to activation inconsistency(which applies for general consensus touching Hardfork), and SDM operation config drift.

The chain will still progress if multiple sequencers use different SDM policy assuming the HF is all correctly activated to all components

Author: Agreed. I’ll split this into activation inconsistency, which can halt or split the chain, and sequencer production configuration drift, which remains consensus-valid but can cause inconsistent refund behavior. Different valid SDM policies do not prevent chain progress and are better covered by FM2a.

- **Description:** SDM uses the Lagoon activation timestamp across the consensus client, execution client, and
  fault-proof stack. A timestamp mismatch can make one component produce or accept `0x7D` while another considers

>   it inactive. Legacy configuration keys may be silently ignored. An execution client without PostExec support

Reviewer 4: What is "legacy configuration key" ? Searched in the specs but could not find.

Author: Fixed.

  Separately, the operator opt-in defaults off and may differ across sequencer instances, causing refunds to stop
  or vary after a restart. Exposing the mutating admin method can let an unauthorized caller toggle production.
- **Risk Assessment:** High impact, low likelihood with an activation preflight. Cross-component activation
  disagreement can halt or split the chain. An opt-in-only mismatch has lower impact because verification is
  independent of opt-in and missing refunds remain valid.
- **Existing Mitigations:**
  1. The Superchain Registry defines the expected Lagoon activation schedule, and every component derives SDM
     activation from that schedule.
  2. Refund production is enabled only when both Lagoon is active and the operator opt-in is set.
  3. Verification ignores the local production opt-in.
  4. The operator opt-in defaults off, supports boot-time configuration, and exposes effective state through the
     admin status method.
- **Possible Mitigations:**
  1. Add a cross-component activation preflight covering timestamps, legacy keys, execution-client support, and
     proof-program selection.

>   2. Add restart health checks that verify the intended effective opt-in state.

Reviewer 4: Lets say there were three sequencers, and all lagoon hf activated. For some reason, single sequencer was opt-out, not producing postExec blocks, and other two were opt-in. We should detect this drift and consolidate the behavior.

Author: Agreed. We should compare the effective SDM opt-in status across all active sequencers and alert on differences, especially after restarts. Monitoring PostExec production can supplement this, although absence of `0x7D` alone is not conclusive because a block may have no eligible refunds. I’ll add this detection and consolidation step to FM5b.

is an explicit sequencer trust assumption.

SDM activates with the Lagoon network upgrade. Producing refunds additionally requires an operator-controlled
execution-layer opt-in. Verification does not depend on this opt-in: after activation, every verifier must accept
and apply a valid SDM payload if the sequencer did.

The normative specifications are the source of truth:

- [Post-Execution Transactions](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/post-exec.md)
- [Sequencer-Defined Metering](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/sdm.md)
- [Lagoon Network Upgrade](https://github.com/ethereum-optimism/specs/blob/main/specs/protocol/lagoon/overview.md)
- [SDM design document](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/sdm.md)

## Failure Modes and Recovery Paths

### FM1a: Structurally invalid or inconsistent PostExec blocks

- **Description:** A producer may seal a malformed PostExec block whether it uses standard block building or
  incremental subblocks. Examples include multiple `0x7D` transactions, an incorrect position or block anchor, and

>   invalid transaction indices. A batcher may omit, corrupt, or reorder `0x7D`; omission can produce a valid derived

Reviewer: > A batcher may omit, corrupt, or reorder `0x7D`

Does it make sense for a batcher to decode (ideally using the exact same lib as derivation) the batch that it just encoded and verify that it matches as a verification step?

Author: I have thought about the same, but I think that'd explode the complexity in the batcher.

  block without refunds that differs from the unsafe block, while corruption or reordering can make the derived
  payload invalid. With subblock streaming enabled, a producer may additionally emit a stale cumulative PostExec
  payload or a final subblock-stream payload that differs from the sealed transaction. A subblock-stream-only mismatch

>   breaks pending views and preconfirmations. Uniform rejection of a malformed sealed block causes a deposits-only

Reviewer: > With subblock streaming enabled, a producer may additionally emit a stale cumulative PostExec payload or a final subblock-stream payload that differs from the sealed transaction. A subblock-stream-only mismatch breaks pending views and preconfirmations.

Is there a reason why the `0x7D` must be included in subblocks? It introduces this necessity for replace-not-append, and adds some failure modes, but I'm unsure what it actually does for the consumer of a subblocks stream. If my understanding is correct, the contract for subblocks from a consumer's standpoint is that any observed transactions are guaranteed to be in that order in the sealed block, but the subblocks tx stream is not guaranteed to be complete (i.e. there may be more txs in the finalized block _after_ the txs that the consumer has read from the websocket). 

Is there a world in which it could be omitted from subblocks entirely and just appended to the tx list when the block is built and returned to the CL?

Author: > Is there a reason why the 0x7D must be included in subblocks?

Every `subblock` must produce a valid block, so yes.

- **Possible Detection:** Alerts for deposits-only replacements.
- **Recovery Path(s):** A subblock-stream-only mismatch self-heals when the canonical block arrives but invalidates
  the affected preconfirmation. Batcher omission can reorg the unsafe block to a valid derived block without refunds.
  Uniform rejection of a malformed sealed or batched payload results in a deposits-only block and requires users to
  resubmit omitted transactions.

### FM1b: PostExec application diverges across clients

- **Description:** Native op-reth execution and kona-executor's fault-proof execution path may apply the same valid
  PostExec payload differently. Examples include applying a refund twice or calculating canonical gas, receipts, or
  settlement differently. A validity, block-hash, or state-root disagreement between the two paths causes a chain
  split or an invalid fault-proof result.
- **Risk Assessment:** High impact, low likelihood. PostExec affects consensus-visible gas accounting, receipts,
  balances, and state roots. The two paths share core execution code, reducing the likelihood, but their parsing,

>   configuration, and integration paths can still diverge.

Reviewer: Re: low likelihood, applying a refund twice was already a bug, right? Just confirming the mitigations that have been added since then to move this to low likelihood.

Author: Within a specific policy - yes, not within SDM.

- **Existing Mitigations:**
  1. Entries may target only standard Ethereum transactions. Execution enforces `refund <= evmGasUsed`, rejects
     settlement underflow, and consumes every entry exactly once.
  2. Existing tests cover producer/verifier round trips, settlement, and structural validity.
- **Possible Mitigations:**
  1. Add op-reth and kona-executor parity coverage for a block containing an SDM refund.

>   2. Add differential coverage across native and fault-proof execution for valid and malformed PostExec payloads.

Reviewer: These are desirable additions, as the "Existing Mitigations" don't seem like they are actually mitigations in that they help define validity, but the FM is that a valid PostExec payload is _applied_ differently.

Author: Agree, we should add tests for these, they are in progress.

  intended policy. An over-refund reduces fee-vault or sequencer revenue; an under-refund overcharges users; a
  selectively applied policy may violate operator commitments. Verifiers cannot distinguish an intended refund
  from an erroneous one because refund policy is not part of consensus. Historically, this class included
  rebates for intrinsically warm or otherwise unmetered accesses: the transaction received a cold-to-warm rebate
  even though the EVM never charged the cold surcharge.
- **Risk Assessment:** Medium impact, low likelihood. The economic misallocation is bounded per transaction by its
  `evmGasUsed` and does not cause a chain split.
- **Existing Mitigations:**
  1. Consensus limits each refund to the transaction's `evmGasUsed`.
  2. Policy tests cover EIP-2929 metered accesses and known intrinsically warm or protocol-generated accesses that
     must not receive a rebate.
  3. Stateful policy data is snapshotted and restored when a candidate transaction fails execution or is not
     committed to the block.
  4. Operators can disable refund production without changing verification behavior.

> - **Possible Mitigations:**

Reviewer: Could we add auditing to live code as a measure of verification? For the block warming example, something like:
- Keep a map of slot -> access count
- Calculate the total expected rebate for the block as `(access_count-1) × 2000` for each slot
- Verify that the cumulative 0x7D refund matches the separately-calculated value

Note: may need to be more granular than just "access" to differentiate between SLOAD, SSTORE, etc. and change `2000` to be the price of the op in question. This may become more difficult, but the general question of auditing via a separate but valid calc holds.

Author: We have that in the `replay` crate. It really doesn't belong on the verifier side though, as we are building SDM to support any kind of policy, not just deterministic ones.

- **Existing Detection:** Refund totals and fee-vault balance changes are observable through metrics and
  canonical block data.
- **Possible Detection:** Alert on refund-to-raw-gas anomalies and compare sampled payloads with an independent
  policy-aware checker.
- **Recovery Path(s):** Disable refund production, correct the policy, and redeploy. Already finalized refunds are
  consensus-valid and should not be rolled back; material losses require accounting and, if appropriate,
  treasury reconciliation.

### FM2b: Settlement violates value conservation

- **Description:** A settlement implementation bug may credit the sender without applying equal debits, debit the
  wrong fee recipient, or otherwise violate value conservation. If deployed consistently across clients, this
  could mint ETH or drain fee-vault or beneficiary balances without causing an immediate chain split.
- **Risk Assessment:** Very high impact, very low likelihood. A value-conservation failure can corrupt canonical
  balances, but the settlement equations, underflow checks, and conservation tests substantially reduce the
  likelihood.
- **Existing Mitigations:**

>   1. Settlement requires sender credits to equal the sum of recipient debits and rejects debit underflow.

Reviewer: How is this enforced? Getting into implementation-specific details, is this something that the typestate pattern in Rust could mitigate and be easily unit tested?

Author: Search for `test_post_exec_settlement_*` - on a unit test level.

Author: If you're asking how this is implemented see `execute_transaction_without_commit`.

We first compute deltas at `post_exec_settlement_deltas` for a given tx.

We then apply the delta for the tx at `apply_post_exec_refund_to_state`.
